### PR TITLE
Draft BVH Refactoring

### PR DESCRIPTION
**Pull request by JuanGonzalezCaminero**
* Adapting the BVHNavigator to work with the modified BVH in VecGeom
* Validation done:
  * Example17, 1 event, 1000 e-, Gun 10-170 in Theta, 0-360 in Phi: Numerically identical to master
  * TestEM3: Numerically identical to master
* Performance comparison:
  * Example17, 64 events, 1000 e-, Gun 10-170 in Theta, 0-360 in Phi.
  * Average Event time ratio new/master : 1.00081
  * Average ECAL time ratio new/master : 1.00423
  * Average Non EM time ratio new/master: 0.99945

### CODE REVIEW

File: base/inc/AdePT/BVHNavigator.h [hunk 1]
```diff
@@ -104,7 +170,17 @@ class BVHNavigator {
 
     if (pvol->GetDaughters().size() > 0) {
       auto bvh = vecgeom::BVHManager::GetBVH(pvol->GetLogicalVolume()->id());
-      bvh->CheckDaughterIntersections(localpoint, localdir, step, pvol, hitcandidate);
```
Comment (hahnjo) [inline base/inc/AdePT/BVHNavigator.h:107]: Can you please fix the issue of not passing the correct `last_exited` in a separate PR that we can validate on its own?

File: base/inc/AdePT/BVHNavigator.h [hunk 2]
```diff
@@ -140,13 +216,19 @@ class BVHNavigator {
...
-    Precision step          = step_limit;
-    VPlacedVolumePtr_t pvol = in_state.Top();
+    Precision step                 = step_limit;
+    VPlacedVolumePtr_t pvol        = in_state.Top();
+    VPlacedVolumePtr_t last_exited = in_state.GetLastExited();
 
     if (pvol->GetDaughters().size() > 0) {
       auto bvh = vecgeom::BVHManager::GetBVH(pvol->GetLogicalVolume()->id());
-      // bvh->CheckDaughterIntersections(localpoint, localdir, step, pvol, hitcandidate);
-      bvh->ApproachNextDaughter(localpoint, localdir, step, pvol);
```
Comment (hahnjo) [inline base/inc/AdePT/BVHNavigator.h:149]: same here, I guess

File: base/inc/AdePT/BVHNavigator.h [hunk 3]
```diff
@@ -31,6 +33,56 @@ class BVHNavigator {
...
+  {
+    return vecgeom::globaldevicegeomdata::gDeviceLogicalVolumes[aLVIndex].GetDaughters()[index];
+  }
+
+  __host__ __device__ static inline vecgeom::VPlacedVolume *GetPlacedVolume(int global_index)
+  {
+#ifdef VECCORE_CUDA_DEVICE_COMPILATION
+    return &vecgeom::globaldevicegeomdata::gCompactPlacedVolBuffer[global_index];
+#else
+    return vecgeom::GeoManager::Instance().FindPlacedVolume(global_index);
```
Comment (agheata) [inline base/inc/AdePT/BVHNavigator.h:46]: See my comment about using direct access to the volume by id in https://gitlab.cern.ch/VecGeom/VecGeom/-/merge_requests/999#note_7095384

File: base/inc/AdePT/BVHNavigator.h [hunk 4]
```diff
@@ -44,13 +96,24 @@ class BVHNavigator {
...
       auto bvh = vecgeom::BVHManager::GetBVH(v->GetLogicalVolume()->id());
 
-      if (!bvh->LevelLocate(exclude, currentpoint, v, daughterlocalpoint)) break;
+      exclude_id = -1;
+      if (exclude != nullptr) {
+        exclude_id = exclude->id();
+      }
+      vol_id = -1;
+
+      if (!bvh->LevelLocate<CandidateContains, SkipItem, ItemId>(exclude_id, currentpoint, vol_id, daughterlocalpoint))
```
Comment (agheata) [inline base/inc/AdePT/BVHNavigator.h:111]: See my comment about the template mechanism: https://gitlab.cern.ch/VecGeom/VecGeom/-/merge_requests/999#note_7095383

**Review by agheata — CHANGES_REQUESTED** [review]

File: base/inc/AdePT/BVHNavigator.h (hunk 1 above)
Comment (JuanGonzalezCaminero) [reply base/inc/AdePT/BVHNavigator.h:107]: https://github.com/apt-sim/AdePT/pull/262

**Comment by agheata** [comment]
This PR was overseeded by https://github.com/apt-sim/AdePT/pull/264

**Pull request closed without merging**